### DISCUSSION TITLE
[v5] condence API to remove /actions

### DISCUSSION THREAD

**Discussion by tomByrer** [statelyai/xstate#1831] [Ideas] [answered]
To me doing imports like this is cumbersome & frustrating for newbies:
```js
import { actions, createMachine, ...} from 'xstate'
const { after, cancel, choose, done, escalate, log, pure, raise, respond, start, stop } = actions
```
I'd much rather not have to copy/paste the entire actions API (I won't remember where is commend is from), & have everything exported directly from xstate:
```js
import { actions, after, cancel, choose, done, escalate, log, pure, raise, respond, start, stop, createMachine, ...} from 'xstate'
```

Maybe this is minor & not worth the change?  Normally I would not complain, but since you're thinking of a v5 & changing some of the API anyhow, perhaps make importing a bit easier?

**Comment by davidkpiano** [comment]
What are your thoughts on using `actions.<actionCreator>`?

```js
import { actions } from 'xstate';

// ...
entry: actions.forwardTo(/* ... */) // etc.
```

Still too cumbersome?

**Comment by Andarist** [comment, marked as answer]
I believe that we should stick to a single way of importing things.

1. either stop exporting individual actions and just leave the `actions` export:
  ```js
  import { actions } from 'xstate'
  actions.assign(/* ... */)
  ```
2. or do the same but move actions to a separate entry:
  ```js
  import * as actions from 'xstate/actions'
  actions.assign(/* ... */)
  
  import { send } from 'xstate/actions'
  send(/* ... */)
  ```
3. or just export everything "flat" from the root entry and remove the `actions` export
  ```js
  import { assign } from 'xstate'
  assign(/* ... */)
  ```

Personally, I would rank them in such order: 2 > 3 > 1

Note that I have also been maintaining a fairly popular project: [redux-saga](https://redux-saga.js.org/) that comes with a very similar `redux-saga/effects` entry point and I don't recall a single issue about users complaining that this is confusing in over 3-4 years.

And also note that we plan to introduce some more entry points (like `xstate/guards`) in v5 so the second option makes the most sense to me as it would not be a single additional entry point but we'd actually have more like this.

**Reply by davidkpiano** [reply]
I think in v5, you're right - 2. seems like the best choice.

**Reply by tomByrer** [reply]
My preference would be 3 > 2 > 1, but the margin between 3 & 2 would be very small.  I'd be OK with your choice, just needs to be **well documented** (eg several times for slow learners like me ;) ).
